dev/tools/gen_keycodes/test/gen_keycodes_test.dart

```diff
@@ -0,0 +1,197 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:gen_keycodes/android_code_gen.dart';
+import 'package:gen_keycodes/base_code_gen.dart';
+import 'package:gen_keycodes/gtk_code_gen.dart';
+import 'package:gen_keycodes/ios_code_gen.dart';
+import 'package:gen_keycodes/logical_key_data.dart';
+import 'package:gen_keycodes/macos_code_gen.dart';
+import 'package:gen_keycodes/physical_key_data.dart';
+import 'package:gen_keycodes/utils.dart';
+import 'package:gen_keycodes/web_code_gen.dart';
+import 'package:gen_keycodes/windows_code_gen.dart';
+import 'package:path/path.dart' as path;
+import 'package:test/test.dart' hide TypeMatcher, isInstanceOf;
+
+String readDataFile(String fileName) {
+  return File(path.join(dataRoot, fileName)).readAsStringSync();
+}
+
+final String dataRoot = path.canonicalize(path.join(Directory.current.absolute.path, 'data'));
+final String testPhysicalData = path.join(dataRoot, 'physical_key_data.json');
+final String testLogicalData = path.join(dataRoot,'logical_key_data.json');
+
+void main() {
+  late String oldDataRoot;
+  setUp(() {
+    oldDataRoot = dataRoot;
+    testDataRoot = dataRoot;
+  });
+
+  tearDown((){
+    testDataRoot = oldDataRoot;
+    oldDataRoot = '';
+  });
+
+  void checkCommonOutput(String output) {
+    expect(output, contains(RegExp('Copyright 201[34]')));
+    expect(output, contains('DO NOT EDIT'));
+    expect(output, contains(RegExp(r'\b[kK]eyA\b')));
+    expect(output, contains(RegExp(r'\b[Dd]igit1\b')));
+    expect(output, contains(RegExp(r'\b[Ff]1\b')));
+    expect(output, contains(RegExp(r'\b[Nn]umpad1\b')));
+    expect(output, contains(RegExp(r'\b[Ss]hiftLeft\b')));
+  }
+
+  test('Generate Keycodes for Android', () {
```

Comment:
These "Generate Keycodes for X" tests seem pretty random and superficial. Are they designed to ensure something that has been broken for you before?

Reply:
I agree, they're not very comprehensive. They are supposed to be smoke tests to make sure that we're generating at least a minimum of the output for each of the platforms without the test being too fragile.

I added them because some symbols I was depending on have disappeared/changed names in the past, and because we didn't have any tests at all for the keygen script. Just having a test file makes it more likely that tests will get written, even if it were just "exit(0)".